### ISSUE TITLE
Cnn

### ISSUE THREAD

**Issue by DevranOz** [DevranOz/devran#2]
from google.colab import drive
drive.mount('/content/drive')

import os
import torch
import torch.nn as nn
import torch.nn.functional as F
from torchvision import datasets, transforms
from torch.utils.data import DataLoader
import matplotlib.pyplot as plt
import numpy as np
from sklearn.metrics import classification_report, confusion_matrix, ConfusionMatrixDisplay, f1_score
from PIL import Image
!pip install grad-cam

from pytorch_grad_cam import GradCAM
from pytorch_grad_cam.utils.image import show_cam_on_image
from pytorch_grad_cam.utils.model_targets import ClassifierOutputTarget

train_dir = "/content/drive/MyDrive/Alz/Train"
test_dir = "/content/drive/MyDrive/Alz/Test"

transform = transforms.Compose([
    transforms.Resize((64, 64)),
    transforms.ToTensor(),
])

train_data = datasets.ImageFolder(train_dir, transform=transform)
test_data = datasets.ImageFolder(test_dir, transform=transform)
train_loader = DataLoader(train_data, batch_size=32, shuffle=True)
test_loader = DataLoader(test_data, batch_size=32, shuffle=False)

class AlzheimerCNN(nn.Module):
    def __init__(self):
        super(AlzheimerCNN, self).__init__()
        self.conv1 = nn.Conv2d(3, 16, kernel_size=3, padding=1)
        self.conv2 = nn.Conv2d(16, 32, kernel_size=3, padding=1)
        self.pool = nn.MaxPool2d(2, 2)
        self.fc1 = nn.Linear(32 * 16 * 16, 128)
        self.fc2 = nn.Linear(128, 4)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))  # [B, 16, 32, 32]
        x = self.pool(F.relu(self.conv2(x)))  # [B, 32, 16, 16]
        x = x.view(-1, 32 * 16 * 16)
        x = F.relu(self.fc1(x))
        return self.fc2(x)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = AlzheimerCNN().to(device)

criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

for epoch in range(40):
    model.train()
    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)
        optimizer.zero_grad()
        loss = criterion(model(images), labels)
        loss.backward()
        optimizer.step()

model.eval()
all_preds = []
all_labels = []

with torch.no_grad():
    for images, labels in test_loader:
        images, labels = images.to(device), labels.to(device)
        outputs = model(images)
        _, preds = outputs.max(1)
        all_preds.extend(preds.cpu().numpy())
        all_labels.extend(labels.cpu().numpy())

class_names = train_data.classes

print("Classification Report:\n", classification_report(all_labels, all_preds, target_names=class_names))
print("Weighted F1 Score:", f1_score(all_labels, all_preds, average="weighted"))

cm = confusion_matrix(all_labels, all_preds)
disp = ConfusionMatrixDisplay(confusion_matrix=cm, display_labels=class_names)
disp.plot(cmap="Blues")
plt.title("Confusion Matrix (Alzheimer CNN)")
plt.show()

image, label = test_data[0] 
input_tensor = image.unsqueeze(0).to(device)

target_layer = model.conv2

cam = GradCAM(model=model, target_layers=[target_layer])
grayscale_cam = cam(input_tensor=input_tensor)[0, :]

rgb_image = image.permute(1, 2, 0).numpy()
rgb_image = (rgb_image - rgb_image.min()) / (rgb_image.max() - rgb_image.min())
visualization = show_cam_on_image(rgb_image, grayscale_cam, use_rgb=True)

plt.imshow(visualization)
plt.title("Grad-CAM Visualization (Alzheimer - CNN)")
plt.axis("off")
plt.show()

5 epoch

![Image](https://github.com/user-attachments/assets/8d0a9081-a92e-42b7-b7e2-3c8d2f4ecc72)
![Image](https://github.com/user-attachments/assets/d3948fcd-a07a-410b-9845-3f9936993eeb)

20 epoch

![Image](https://github.com/user-attachments/assets/1db06a73-e019-41e3-8b39-da3f9a3b0f74)
![Image](https://github.com/user-attachments/assets/9939bde8-a57f-4597-99e1-d2e2d71ba8bd)

40 epoch

![Image](https://github.com/user-attachments/assets/973af627-be69-47ac-988f-81f75cd462f7)
![Image](https://github.com/user-attachments/assets/d0b3e1ca-0dd3-4a21-8cfe-ff42c69df4a2)
![Image](https://github.com/user-attachments/assets/b65fd965-0152-42f0-87f8-9b7a96bae9d4)
